src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -57,11 +107,34 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
             _clients = clients;
         }
 
-        public IEnumerable<ConfigurationClient> GetAvailableClients(DateTimeOffset time)
+        public async Task<IEnumerable<ConfigurationClient>> GetAvailableClients(DateTimeOffset time, CancellationToken cancellationToken)
         {
-            return _clients.Where(client => client.BackoffEndTime <= time).Select(c => c.Client).ToList();
-        }
+            if (_replicaDiscoveryEnabled)
+            {
+                // The source of truth whether there's avaliable client in the list.
+                int successfulClientsCount = _clients.Where(client => client.FailedAttempts == 0).Count();
+
+                // Refresh fallback clients if:
+                // 1.It's the initial attempt to query available clients on app start-up.
+                // 2.At least one client is available in the last attempt, it's the first attempt that all static and dynamic clients are backed off.
+                // 3.It's not the first attempt that all clients have backed off, as well as it's due time to refresh fallback clients.
+
+                if ((_lastFallbackClientRefresh == default ||
+                    successfulClientsCount == 0) &&
+                    !_allClientFailed ||
+                    time >= _lastFallbackClientRefresh + _fallbackClientRefreshInterval)

```

Comment:
There's an `&&` without a paranthesis. Is that intentional?

Also, I recommend that we have to separate refresh intervals.

A healthy interval ~5 min just so we stay up to date, and an unhealthy interval ~30s that is lower but ensures that we don't cause to much noise if we hit a condition where we want to detect a newly created replica.

Reply:
It's intentional.

If it's healthy status (at least one client can get a successful return code), I think we don't need to bother to perform a lookup, it's OK even if we don't stay up to date for a long time, it' not late to refresh the status immediately once it fall into unhealthy. That's why I introduce `_allClientFailed`.

I agree with an unhealthy interval ~30s, ensure it can quickly be refreshed with newly created replica(s)

Reply:
> If it's healthy status (at least one client can get a successful return code), I think we don't need to bother to perform a lookup

For highest availability optimistically refresh caches in the event that the dependent service (srv records) are not available when you need them.